crates/networking/manager/src/gossipsub/validate/beacon_block.rs

```
@@ -0,0 +1,275 @@
+                address: validator.public_key.clone(),
+                slot: block.message.slot,
+            },
+            signed_proposer_bls_execution_change.message.clone(),
+        );
+
+    Ok(ValidationResult::Accept)
+}
+
+pub async fn validate_beacon_block(
+    beacon_chain: &BeaconChain,
+    cached_db: &CachedDB,
+    block: &SignedBeaconBlock,
+    state: &BeaconState,
+    is_parent: bool,
+) -> anyhow::Result<ValidationResult> {
+    let store = beacon_chain.store.lock().await;
+
+    let current_global_slot = store.get_current_slot()?;
+
+    // [IGNORE] The block is not from a future slot.
+    if block.message.slot > current_global_slot {
+        return Ok(ValidationResult::Ignore(
+            "Block is from a future slot".to_string(),
+        ));
+    }
+
+    // [IGNORE] The block is from a slot greater than the latest finalized slot.
+    if block.message.slot
+        <= compute_start_slot_at_epoch(store.db.finalized_checkpoint_provider().get()?.epoch)
+    {
+        return Ok(ValidationResult::Ignore(
+            "Block is from a slot greater than the latest finalized slot".to_string(),
+        ));
+    }
+
+    let validator = state
+        .validators
+        .get(block.message.proposer_index as usize)
+        .ok_or(anyhow!("Invalid proposer index"))?;
+
+    // [IGNORE] The block is the first block with valid signature received for the proposer for the
+    // slot.
+    if cached_db
+        .cached_proposer_signature
+        .read()
+        .await
+        .contains(&AddressSlotIdentifier {
+            address: validator.public_key.clone(),
+            slot: block.message.slot,
+        })
+    {
+        return Ok(ValidationResult::Ignore(
+            "Signature already received".to_string(),
+        ));
+    }
+
+    // [REJECT] The proposer signature, signed_beacon_block.signature, is valid with respect to the
+    // proposer_index pubkey.
+    if !state.verify_block_signature(block)? {
```

From [`verify_block_signature()` code](https://github.com/ReamLabs/ream/blob/master/crates/common/consensus/src/electra/beacon_state.rs#L2409-L2420):

```rs
    pub fn verify_block_signature(&self, signed_block: &SignedBeaconBlock) -> anyhow::Result<bool> {
        let proposer = &self.validators[signed_block.message.proposer_index as usize];
        let signing_root = compute_signing_root(
            signed_block.message.clone(),
            self.get_domain(DOMAIN_BEACON_PROPOSER, None),
        );

        signed_block
            .signature
            .verify(&proposer.public_key, signing_root.as_ref())
            .map_err(|err| anyhow!("Invalid block signature: {err}"))
    }
```

By using `?`, isn't an `Err` of invalid block signature early-returned out of `validate_beacon_block()` immediately (which is behaves closer to an ignore) instead of returning `ValidationResult::Reject`?

---

Hmm but now with https://github.com/ReamLabs/ream/pull/618/commits/2a369c5676f5b2f17a4ee62b551993df5c34a3c2, wouldn't all other uses of `?` result in a reject?

E.g. `store.db.finalized_checkpoint_provider().get()?`, `get(block.message.parent_root)?`, etc. where they are internal errors, while previously it would have gone into this logic: https://github.com/ReamLabs/ream/pull/618/files#diff-5311d8512b40fed4001b0c59e6cc2f0c1369832d53984d0fe4a41eee4e34d7bdR104-R107

```rs
let validation_result = match validate_gossip_beacon_block(
    beacon_chain,
    cached_db,
    &signed_block,
)
.await
{
    Ok(result) => result,
    Err(err) => { // <-- This ---
        warn!("Failed to validate gossipsub beacon block: {err}");
        return;
    }
};
```

---

I'll resolve this one but I have one follow up suggestion: https://github.com/ReamLabs/ream/pull/618#pullrequestreview-3000516187